pkg/controller/v1alpha1/llmisvc/workload.go

```diff
@@ -22,51 +22,118 @@ import (
 	logger := log.FromContext(ctx).WithName("reconcileWorkload")
 	ctx = log.IntoContext(ctx, logger)
 
 	logger.Info("Reconciling Workload")
 
+	// Ensure readiness is determined even if errors occur
 	defer llmSvc.DetermineWorkloadReadiness()
 
+	// Set up TLS certificates for secure communication
 	if err := r.reconcileSelfSignedCertsSecret(ctx, llmSvc); err != nil {
+		llmSvc.MarkMainWorkloadNotReady("ReconcileCertsError", err.Error())
 		return fmt.Errorf("failed to reconcile self-signed certificates secret: %w", err)
 	}
 
 	// We need to always reconcile every type of workload to handle transitions from P/D to another topology (meaning
 	// finalizing superfluous workloads).
 
-	if err := r.reconcileMultiNodeWorkload(ctx, llmSvc); err != nil {
-		llmSvc.MarkWorkloadNotReady("ReconcileMultiNodeWorkloadError", err.Error())
+	// Handle multi-node deployments using LeaderWorkerSets
+	if err := r.reconcileMultiNodeWorkload(ctx, llmSvc, storageConfig, credentialConfig); err != nil {
+		llmSvc.MarkMainWorkloadNotReady("ReconcileMultiNodeWorkloadError", err.Error())
 		return fmt.Errorf("failed to reconcile multi node workload: %w", err)
 	}
 
-	if err := r.reconcileSingleNodeWorkload(ctx, llmSvc, storageConfig); err != nil {
-		llmSvc.MarkWorkloadNotReady("ReconcileSingleNodeWorkloadError", err.Error())
+	// Handle single-node deployments using standard Deployments
+	if err := r.reconcileSingleNodeWorkload(ctx, llmSvc, storageConfig, credentialConfig); err != nil {
+		llmSvc.MarkMainWorkloadNotReady("ReconcileSingleNodeWorkloadError", err.Error())
 		return fmt.Errorf("failed to reconcile single node workload: %w", err)
 	}
 
+	// Create Service to expose workload pods
+	if err := r.reconcileWorkloadService(ctx, llmSvc); err != nil {
+		llmSvc.MarkMainWorkloadNotReady("ReconcileWorkloadServiceError", err.Error())
+		return fmt.Errorf("failed to reconcile workload service: %w", err)
+	}
+
 	return nil
 }
 
+func (r *LLMISVCReconciler) reconcileWorkloadService(ctx context.Context, llmSvc *v1alpha1.LLMInferenceService) error {
+	expected := &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      kmeta.ChildName(llmSvc.GetName(), "-kserve-workload-svc"),
+			Namespace: llmSvc.GetNamespace(),
+			Labels: map[string]string{
+				"app.kubernetes.io/component": "llminferenceservice-workload",
+				"app.kubernetes.io/name":      llmSvc.GetName(),
+				"app.kubernetes.io/part-of":   "llminferenceservice",
+			},
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(llmSvc, v1alpha1.LLMInferenceServiceGVK),
+			},
+		},
+		Spec: corev1.ServiceSpec{
+			Ports: []corev1.ServicePort{
+				{
+					Name:        "https",
+					Protocol:    corev1.ProtocolTCP,
+					AppProtocol: ptr.To("https"),
+					Port:        8000,
+					TargetPort: intstr.IntOrString{
+						Type:   intstr.Int,
+						IntVal: 8000,
```

Comment:
why this is hard coded ? it depends on the workload ports defined on container spec

Reply:
Added